# Preprocessing

In [1]:
import pandas as pd
import numpy as np

In [2]:
from functions import *

### Import the datasets

In [3]:
# Open the raw datasets
# Heavy Metal
hm_df = pd.read_excel('data/AllHMFiles.xlsx', engine='openpyxl').fillna(0.0)

# Pesticides
pesticides_df = pd.read_excel('data/AllPesticidesFiles.xlsx', engine='openpyxl').fillna(0.0)
classe_pesticides_df = pd.read_excel('data/pesticides.xlsx', engine='openpyxl')

# Surfaces
surfaces_df = pd.read_excel('data/absSurfs.xlsx', engine='openpyxl')

# Distances
distances_df = pd.read_excel('data/distsOneSheet.xlsx', engine='openpyxl')

In [4]:
hm_df.head(5)

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  Zinc  Chrome  
0     0.032    40     0.0  
1     0.029    33     0.0  
2     0.038    37     0.0  
3     0.031    37     0.0  
4     0.052    35     0.0

In [5]:
pesticides_df.head(5)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
0            B0017  2017/P1     0.0     0.0000          0.00           0.0   
1            B0019  2017/P1     0.0     0.0188          0.00           0.0   
2            B0021  2017/P1     0.0     0.0711          0.00           0.0   
3            B0023  2017/P1     0.0     0.0294          0.00           0.0   
4            B0037  2017/P1     0.0     0.0000          0.01           0.0   

   Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  Nicosulfuron  \
0           0.0           0.0          0.0        0.0  ...           0.0   
1           0.0           0.0          0.0        0.0  ...           0.0   
2           0.0           0.0          0.0        0.0  ...           0.0   
3           0.0           0.0          0.0        0.0  ...           0.0   
4           0.0           0.0          0.0        0.0  ...           0.0   

   Piperonylbutoxide  Prosulfuron  Simazine  Terbuthylazine  Dieldrin  \
0                0.0          0.0       0.0             0.0       0.0   
1                0.0          0.0       0.0             0.0       0.0   
2                0.0          0.0       0.0             0.0       0.0   
3                0.0          0.0       0.0             0.0       0.0   
4                0.0          0.0       0.0             0.0       0.0   

   Fenamidone  Trifluralin  Cycloate  Pentachloroanisol  
0         0.0          0.0       0.0                0.0  
1         0.0          0.0       0.0                0.0  
2         0.0          0.0       0.0                0.0  
3         0.0          0.0       0.0                0.0  
4         0.0          0.0       0.0                0.0  

[5 rows x 117 columns]

In [6]:
classe_pesticides_df.head(5)

importName           name    statusEN   LMR  \
0      Aclonifen      aclonifen  authorized  0.05   
1         Aldrin         aldrin  prohibited  0.01   
2   Ametoctradin   ametoctradin  authorized  0.05   
3  Anthraquinone  anthraquinone  prohibited  0.02   
4   Azoxystrobin   azoxystrobin  authorized  0.05   

                          familyEN          typeEN  
0                   diphenyl ether       herbicide  
1                   organochlorine     insecticide  
2               triazolopyrimidine       fungicide  
3  polycyclic aromatic hydrocarbon  bird repellant  
4                      strobilurin       fungicide

In [7]:
surfaces_df.head(5)

Site  111           112           121            122  123  124  131  132  \
0  F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1  F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2  F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3  B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4  F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   

   133  ...  411  412  421  422  423  511            512  521  522  523  
0  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4  0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  

[5 rows x 45 columns]

In [8]:
distances_df.head(5)

Site     polyID  classCLC         dist
0  F0069   FR-88090       231  1183.022362
1  F0069  FR-131773       242   289.926324
2  F0069   FR-46167       211   221.344153
3  F0069   FR-11047       112  1038.383069
4  F0069  FR-195343       311   724.045683

### Heavy Metal

In [9]:
hm_df.head()

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  Zinc  Chrome  
0     0.032    40     0.0  
1     0.029    33     0.0  
2     0.038    37     0.0  
3     0.031    37     0.0  
4     0.052    35     0.0

In [10]:
MRL_HM_dict = {
    'Lead' : [1.5],
    'Cadmium' : [1],
    'Arsenic' : [0.3],
    'Mercury' : [0],
    'Copper' : [30]
}

MRL_HM = pd.DataFrame.from_dict(MRL_HM_dict)
MRL_HM

Lead  Cadmium  Arsenic  Mercury  Copper
0   1.5        1      0.3        0      30

In [11]:
hm_df['mrl_lead'] = np.where(hm_df['Lead'] > 1.5, 1, 0)
hm_df['mrl_cadmium'] = np.where(hm_df['Cadmium'] > 1, 1, 0)
hm_df['mrl_arsenic'] = np.where(hm_df['Arsenic'] > 0.3, 1, 0)
hm_df['mrl_copper'] = np.where(hm_df['Copper'] > 30, 1, 0)
hm_df['has_mercury'] = np.where(hm_df['Mercury'] > 0, 1, 0)

hm_df.head()

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  Zinc  Chrome  mrl_lead  mrl_cadmium  mrl_arsenic  mrl_copper  \
0     0.032    40     0.0         0            0            0           0   
1     0.029    33     0.0         0            0            0           0   
2     0.038    37     0.0         0            0            0           0   
3     0.031    37     0.0         0            0            0           0   
4     0.052    35     0.0         0            0            0           0   

   has_mercury  
0            0  
1            0  
2            0  
3            0  
4            0

In [12]:
# Dropping columns of metals without specific threshhold level defined by laws
hm_df = hm_df.drop(['Chromium'], axis=1)
hm_df = hm_df.drop(['Chrome'], axis=1)
hm_df = hm_df.drop(['Zinc'], axis=1)

# Dropping other columns not needed anymore
hm_df = hm_df.drop(['Lead'], axis=1)
hm_df = hm_df.drop(['Cadmium'], axis=1)
hm_df = hm_df.drop(['Arsenic'], axis=1)
hm_df = hm_df.drop(['Mercury'], axis=1)
hm_df = hm_df.drop(['Copper'], axis=1)

hm_df.head()

REF....SUBSTANCE   PERIOD  mrl_lead  mrl_cadmium  mrl_arsenic  mrl_copper  \
0            B0017  2017/P1         0            0            0           0   
1            B0019  2017/P1         0            0            0           0   
2            B0021  2017/P1         0            0            0           0   
3            B0023  2017/P1         0            0            0           0   
4            B0037  2017/P1         0            0            0           0   

   has_mercury  
0            0  
1            0  
2            0  
3            0  
4            0

##### Aleternate dataset (grooup by) for HM

In [13]:
group_hm_df = hm_df.groupby('REF....SUBSTANCE').agg(max).drop(columns=['PERIOD'])
group_hm_df.head()

mrl_lead  mrl_cadmium  mrl_arsenic  mrl_copper  has_mercury
REF....SUBSTANCE                                                             
B0012                    0            0            0           0            0
B0017                    0            0            0           0            0
B0019                    0            0            0           0            1
B0021                    0            1            0           0            0
B0023                    0            1            0           0            0

##### Alternate df, with categorical

In [14]:
cat_hm_df = group_hm_df.copy(deep=True)

In [15]:
cat_hm_df['cat'] = cat_hm_df.apply(lambda x: (str(x['mrl_lead']) + 
                                              str(x['mrl_cadmium']) + 
                                              str(x['mrl_arsenic']) + 
                                              str(x['mrl_copper'])) + 
                                              str(x['has_mercury']), axis=1)

In [16]:
cat_hm_df['cat'] = cat_hm_df['cat'].astype('category')
cat_hm_df['cat_label'] = cat_hm_df['cat'].cat.codes

labels = cat_hm_df.set_index('cat_label')['cat'].to_dict()
labels

{0: '00000',
 1: '00001',
 4: '01000',
 5: '10000',
 3: '00100',
 6: '10001',
 2: '00010'}

In [17]:
cat_hm_df = cat_hm_df.drop(columns=['mrl_lead', 
                                    'mrl_cadmium', 
                                    'mrl_arsenic', 
                                    'mrl_copper',
                                    'has_mercury',
                                    'cat'])

In [18]:
cat_hm_df.sample(5)

cat_label
REF....SUBSTANCE           
F0073                     0
F0057                     0
F0029                     0
B0158                     4
F0005                     0

### Pesticides

Fungicide :

In [19]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'fungicide'].head(5)

importName                       name    statusEN   LMR  \
2         Ametoctradin               ametoctradin  authorized  0.05   
4         Azoxystrobin               azoxystrobin  authorized  0.05   
5            Benalaxyl                  benalaxyl  prohibited  0.05   
7  Benthiavalicarb.iso  benthiavalicarb-isopropyl  authorized  0.05   
8     Benzovindiflupyr           benzovindiflupyr  authorized  0.05   

             familyEN     typeEN  
2  triazolopyrimidine  fungicide  
4         strobilurin  fungicide  
5        phenylamides  fungicide  
7           carbamate  fungicide  
8               amide  fungicide

In [20]:
fungicides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'fungicide'].importName.to_list()
fungicides.append('Diphenylamine')
fungicides.append('Tebuconazole')
fungicides

['Ametoctradin',
 'Azoxystrobin',
 'Benalaxyl',
 'Benthiavalicarb.iso',
 'Benzovindiflupyr',
 'Boscalid',
 'Bupirimate',
 'Captan',
 'Carbendazim',
 'Chlorothalonil',
 'Cyazofamid',
 'Cyflufenamid',
 'Cymoxanil',
 'Cyprodinil',
 'Difenoconazole',
 'Dimethomorph',
 'Dimoxystrobin',
 'Dodine',
 'Epoxiconazole',
 'Ethirimol',
 'Famoxadone',
 'Fenamidone',
 'Fenbuconazole',
 'Fenhexamid',
 'Fenpropidin',
 'Fenpropimorph',
 'Fluazinam',
 'Fludioxonil',
 'Fluopicolide',
 'Fluopyram',
 'Flutolanil',
 'Fluxapyroxad',
 'Hexaconazole',
 'Iprovalicarb',
 'Isofetamid',
 'Mandipropamid',
 'Metalaxyl',
 'Metconazole',
 'Metrafenone',
 'Penconazole',
 'Pencycuron',
 'Pentachloroanisol',
 'Prochloraz',
 'Propamocarb',
 'Propiconazole',
 'Prothioconazole_d',
 'Pyraclostrobin',
 'Pyrimethanil',
 'Quinoxyfen',
 'Spiroxamine',
 'Tetraconazole',
 'Thiabendazole',
 'Thiophanate.methyl',
 'Trifloxystrobin',
 'X2.phenylphenol',
 'Zoxamide',
 'Diphenylamine',
 'Tebuconazole']

Insecticide :

In [21]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'insecticide'].head(5)

importName                 name          statusEN   LMR  \
1                Aldrin               aldrin        prohibited  0.01   
13  Chlorantraniliprole  chlorantraniliprole        authorized  0.05   
17  Chlorpyrifos.methyl  chlorpyrifos-methyl        prohibited  0.05   
19            Coumaphos            coumaphos        prohibited  0.10   
20         Cyanofenphos         cyanofenphos  not yet assessed   NaN   

               familyEN       typeEN  
1        organochlorine  insecticide  
13  anthranilic diamide  insecticide  
17      organophosphate  insecticide  
19        thiophosphate  insecticide  
20      organophosphate  insecticide

In [22]:
insecticides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'insecticide'].importName.to_list()
insecticides.append('Diphenylamine')
insecticides.append('Methiocarb_total')
insecticides

['Aldrin',
 'Chlorantraniliprole',
 'Chlorpyrifos.methyl',
 'Coumaphos',
 'Cyanofenphos',
 'Cyflumetofen',
 'Cypermethrin',
 'Deet',
 'Dieldrin',
 'Etofenprox',
 'Fenoxycarb',
 'Fenpropathrin',
 'Flonicamid',
 'Flucythrinate',
 'Flupyradifurone',
 'Imidacloprid',
 'Indoxacarb',
 'Methoxyfenozide',
 'Phosmet',
 'Piperonylbutoxide',
 'Pirimicarb',
 'Spinosad',
 'Spirotetramat',
 'Tau.fluvalinate',
 'Tebufenozide',
 'Tetradifon',
 'Thiacloprid',
 'Diphenylamine',
 'Methiocarb_total']

Herbicide : 

In [23]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'herbicide'].head(5)

importName           name    statusEN   LMR          familyEN  \
0       Aclonifen      aclonifen  authorized  0.05    diphenyl ether   
6     Benfluralin    benfluralin  authorized  0.05    dinitroaniline   
15  Chlorotoluron  chlorotoluron  authorized  0.05        phenylurea   
16   Chlorpropham   chlorpropham  prohibited  0.05         carbamate   
18      Clethodim      clethodim  authorized  0.05  cyclohexanedione   

       typeEN  
0   herbicide  
6   herbicide  
15  herbicide  
16  herbicide  
18  herbicide

In [24]:
herbicides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'herbicide'].importName.to_list()
herbicides

['Aclonifen',
 'Benfluralin',
 'Chlorotoluron',
 'Chlorpropham',
 'Clethodim',
 'Cycloate',
 'Diflufenican',
 'Ethofumesate',
 'Fluazifop',
 'Flufenacet',
 'Metamitron',
 'Methabenzthiazuron',
 'Metobromuron',
 'Metolachlor',
 'Nicosulfuron',
 'Pendimethalin',
 'Phenmedipham',
 'Propanil',
 'Propyzamide',
 'Prosulfocarb',
 'Prosulfuron',
 'Simazine',
 'Terbuthylazine',
 'Tri.allate',
 'Trifluralin']

Others :

In [25]:
classe_pesticides_df[(classe_pesticides_df['typeEN'] != 'fungicide') & 
                     (classe_pesticides_df['typeEN'] != 'insecticide') & 
                     (classe_pesticides_df['typeEN'] != 'herbicide')]

importName           name    statusEN   LMR  \
3           Anthraquinone  anthraquinone  prohibited  0.02   
34          Diphenylamine  diphenylamine  prohibited  0.05   
69       Methiocarb_total     methiocarb  prohibited  0.05   
86             Propargite     propargite  prohibited  0.05   
100          Tebuconazole   tebuconazole  authorized  0.05   
112  X6.benzylaminopurine  benzyladenine  authorized   NaN   
114                   NaN            NaN         NaN   NaN   

                            familyEN  \
3    polycyclic aromatic hydrocarbon   
34                             amine   
69                         carbamate   
86                               NaN   
100                         triazole   
112                       cytokinins   
114                              NaN   

                                           typeEN  
3                                  bird repellant  
34    growth regulator, fungicide and insecticide  
69   insecticide, molluscicide and bird repellent  
86                                      acaricide  
100                fungicide and growth regulator  
112                              growth regulator  
114                                           NaN

##### Choice to treat the others

Adding to the corresponding categories, stay in "other" ? What about the growth regulator ?

#### Add some columns

In [26]:
pesticides_df['qtt_fungicides'] = pesticides_df.apply(qtt_pesticides, args = fungicides, axis = 1)
pesticides_df['qtt_insecticides'] = pesticides_df.apply(qtt_pesticides, args = insecticides, axis = 1)
pesticides_df['qtt_herbicides'] = pesticides_df.apply(qtt_pesticides, args = herbicides, axis = 1)

In [27]:
pesticides_df['nb_fungicides'] = pesticides_df.apply(nb_pesticides, args = fungicides, axis = 1)
pesticides_df['nb_insecticides'] = pesticides_df.apply(nb_pesticides, args = insecticides, axis = 1)
pesticides_df['nb_herbicides'] = pesticides_df.apply(nb_pesticides, args = herbicides, axis = 1)

In [28]:
pesticides_df['mrl_fungicides'] = pesticides_df.apply(lmr_pesticides, args = fungicides, axis = 1)
pesticides_df['mrl_insecticides'] = pesticides_df.apply(lmr_pesticides, args = insecticides, axis = 1)
pesticides_df['mrl_herbicides'] = pesticides_df.apply(lmr_pesticides, args = herbicides, axis = 1)

In [29]:
pesticides_df.head(10)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
0            B0017  2017/P1     0.0     0.0000          0.00           0.0   
1            B0019  2017/P1     0.0     0.0188          0.00           0.0   
2            B0021  2017/P1     0.0     0.0711          0.00           0.0   
3            B0023  2017/P1     0.0     0.0294          0.00           0.0   
4            B0037  2017/P1     0.0     0.0000          0.01           0.0   
5            B0039  2017/P1     0.0     0.0000          0.00           0.0   
6            B0041  2017/P1     0.0     0.0000          0.00           0.0   
7            B0043  2017/P1     0.0     0.0000          0.00           0.0   
8            B0045  2017/P1     0.0     0.0000          0.00           0.0   
9            B0059  2017/P1     0.0     0.0000          0.00           0.0   

   Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  Pentachloroanisol  \
0          0.00           0.0         0.00        0.0  ...                0.0   
1          0.00           0.0         0.00        0.0  ...                0.0   
2          0.00           0.0         0.00        0.0  ...                0.0   
3          0.00           0.0         0.00        0.0  ...                0.0   
4          0.00           0.0         0.00        0.0  ...                0.0   
5          0.01           0.0         0.00        0.0  ...                0.0   
6          0.00           0.0         0.00        0.0  ...                0.0   
7          0.00           0.0         0.01        0.0  ...                0.0   
8          0.00           0.0         0.00        0.0  ...                0.0   
9          0.00           0.0         0.00        0.0  ...                0.0   

   qtt_fungicides  qtt_insecticides  qtt_herbicides  nb_fungicides  \
0          0.0000            0.0000            0.00              0   
1          0.0188            0.0000            0.00              1   
2          0.0711            0.0000            0.00              1   
3          0.0294            0.0000            0.00              1   
4          0.0000            0.0000            0.01              0   
5          0.0000            0.0208            0.00              0   
6          0.0000            0.0000            0.00              0   
7          0.0100            0.0000            0.00              1   
8          0.0000            0.0000            0.00              0   
9          0.0145            0.0000            0.00              1   

   nb_insecticides  nb_herbicides  mrl_fungicides  mrl_insecticides  \
0                0              0               0                 0   
1                0              0               0                 0   
2                0              0               1                 0   
3                0              0               0                 0   
4                0              1               0                 0   
5                2              0               0                 0   
6                0              0               0                 0   
7                0              0               0                 0   
8                0              0               0                 0   
9                0              0               0                 0   

   mrl_herbicides  
0               0  
1               0  
2               0  
3               0  
4               0  
5               0  
6               0  
7               0  
8               0  
9               0  

[10 rows x 126 columns]

In [30]:
pesticides_df['year'] = pesticides_df['PERIOD'].apply(lambda x: x[:4])
pesticides_df['trimester'] = pesticides_df['PERIOD'].apply(lambda x: x[-1:])
pesticides_df.sample(10)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
394            B0043  2020/P2     0.0     0.0061           0.0           0.0   
411            F0007  2020/P2     0.0     0.0000           0.0           0.0   
23             B0019  2017/P2     0.0     0.0000           0.0           0.0   
160            B0142  2018/P3     0.0     0.0000           0.0           0.0   
295            B0142  2019/P3     0.0     0.0000           0.0           0.0   
147            B0032  2018/P3     0.0     0.0118           0.0           0.0   
236            B0019  2019/P2     0.0     0.0000           0.0           0.0   
305            F0035  2019/P3     0.0     0.0000           0.0           0.0   
390            B0032  2020/P2     0.0     0.0590           0.0           0.0   
426            F0055  2020/P2     0.0     0.0000           0.0           0.0   

     Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  \
394           0.0           0.0          0.0        0.0  ...   
411           0.0           0.0          0.0        0.0  ...   
23            0.0           0.0          0.0        0.0  ...   
160           0.0           0.0          0.0        0.0  ...   
295           0.0           0.0          0.0        0.0  ...   
147           0.0           0.0          0.0        0.0  ...   
236           0.0           0.0          0.0        0.0  ...   
305           0.0           0.0          0.0        0.0  ...   
390           0.0           0.0          0.0        0.0  ...   
426           0.0           0.0          0.0        0.0  ...   

     qtt_insecticides  qtt_herbicides  nb_fungicides  nb_insecticides  \
394            0.0000          0.0120              2                0   
411            0.0000          0.0000              0                0   
23             0.0000          0.0000              0                0   
160            0.0000          0.0000              0                0   
295            0.0000          0.0000              1                0   
147            0.0000          0.0000              1                0   
236            0.0000          0.0000              0                0   
305            0.0000          0.0000              0                0   
390            0.0000          0.0138              2                0   
426            0.0356          0.0050              2                1   

     nb_herbicides  mrl_fungicides  mrl_insecticides  mrl_herbicides  year  \
394              1               0                 0               0  2020   
411              0               0                 0               0  2020   
23               0               0                 0               0  2017   
160              0               0                 0               0  2018   
295              0               0                 0               0  2019   
147              0               0                 0               0  2018   
236              0               0                 0               0  2019   
305              0               0                 0               0  2019   
390              1               1                 0               0  2020   
426              1               1                 0               0  2020   

     trimester  
394          2  
411          2  
23           2  
160          3  
295          3  
147          3  
236          2  
305          3  
390          2  
426          2  

[10 rows x 128 columns]

##### Alternate (smaller) dataset for pesticides

In [31]:
pesti_names = pesticides_df.columns.to_list()[1:-11]

In [32]:
small_pesticides_df = pesticides_df.drop(pesti_names, axis=1)
small_pesticides_df.head(5)

REF....SUBSTANCE  qtt_fungicides  qtt_insecticides  qtt_herbicides  \
0            B0017          0.0000               0.0            0.00   
1            B0019          0.0188               0.0            0.00   
2            B0021          0.0711               0.0            0.00   
3            B0023          0.0294               0.0            0.00   
4            B0037          0.0000               0.0            0.01   

   nb_fungicides  nb_insecticides  nb_herbicides  mrl_fungicides  \
0              0                0              0               0   
1              1                0              0               0   
2              1                0              0               1   
3              1                0              0               0   
4              0                0              1               0   

   mrl_insecticides  mrl_herbicides  year trimester  
0                 0               0  2017         1  
1                 0               0  2017         1  
2                 0               0  2017         1  
3                 0               0  2017         1  
4                 0               0  2017         1

##### Alternate (group by) dataset for pesticides

In [33]:
group_pesti_df = pesticides_df.groupby('REF....SUBSTANCE').agg(max)

In [34]:
group_pesti_df = group_pesti_df.loc[:,'mrl_fungicides' : 'mrl_herbicides']
group_pesti_df.sample(10)

mrl_fungicides  mrl_insecticides  mrl_herbicides
REF....SUBSTANCE                                                  
B0059                          1                 0               0
FE0001                         0                 0               0
B0093                          1                 0               0
B0140                          0                 0               0
F0020                          1                 0               0
B0152                          1                 0               0
B0103                          1                 0               0
B0032                          1                 0               0
F0011                          0                 1               0
F0013                          0                 0               0

##### Alternate df, with categorical

In [35]:
cat_pesti_df = group_pesti_df.copy(deep=True)

In [36]:
cat_pesti_df['cat'] = cat_pesti_df.apply(lambda x: (str(x['mrl_fungicides']) + 
                                                    str(x['mrl_insecticides']) + 
                                                    str(x['mrl_herbicides']))  , axis=1)

In [37]:
cat_pesti_df.sample(10)

mrl_fungicides  mrl_insecticides  mrl_herbicides  cat
REF....SUBSTANCE                                                       
B0093                          1                 0               0  100
B0163                          1                 0               0  100
F0035                          0                 0               1  001
F0015                          1                 0               0  100
F0005                          0                 0               0  000
F0003                          1                 0               0  100
B0023                          0                 0               0  000
B0019                          0                 0               0  000
B0152                          1                 0               0  100
F0007                          0                 0               0  000

In [38]:
cat_pesti_df['cat'] = cat_pesti_df['cat'].astype('category')
cat_pesti_df['cat_label'] = cat_pesti_df['cat'].cat.codes

In [39]:
labels = cat_pesti_df.set_index('cat_label')['cat'].to_dict()
labels

{3: '100', 0: '000', 5: '110', 4: '101', 6: '111', 2: '010', 1: '001'}

In [40]:
cat_pesti_df = cat_pesti_df.drop(columns=['mrl_fungicides', 
                                          'mrl_insecticides', 
                                          'mrl_herbicides', 
                                          'cat'])

In [41]:
cat_pesti_df.sample(10)

cat_label
REF....SUBSTANCE           
B0067                     3
F0041                     0
F0057                     3
B0116                     0
B0041                     3
B0128                     0
B0093                     3
F0039                     3
B0037                     5
F0001                     5

### Sites

##### Getting the area for each zone

In [42]:
distances_df.head(5)

Site     polyID  classCLC         dist
0  F0069   FR-88090       231  1183.022362
1  F0069  FR-131773       242   289.926324
2  F0069   FR-46167       211   221.344153
3  F0069   FR-11047       112  1038.383069
4  F0069  FR-195343       311   724.045683

In [43]:
surfaces_df.head(5)

Site  111           112           121            122  123  124  131  132  \
0  F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1  F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2  F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3  B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4  F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   

   133  ...  411  412  421  422  423  511            512  521  522  523  
0  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4  0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  

[5 rows x 45 columns]

In [44]:
# surfaces_df['country'] = surfaces_df['Site'].apply(lambda x: 'FR' if x.startswith('F') else 'BE')
# surfaces_df['country'].value_counts()

In [45]:
sites_df = distances_df.copy(deep=True)

In [46]:
sites_df['surface'] = sites_df.apply(get_surface, axis = 1)

In [47]:
sites_df.sample(10)

Site     polyID  classCLC         dist       surface
459  F0039   FR-59392       211     0.000000  4.998301e+06
26   F0027  FR-201572       311   268.999512  3.212519e+06
286  B0093    BE-5694       211  1255.494603  2.350786e+06
481  B0032   BE-14504       313   398.693443  4.353821e+05
521  B0067   BE-14015       242     0.000000  2.383974e+06
305  F0055   FR-32565       131  1035.587970  1.261217e+05
266  B0059    BE-4271       121   694.110957  3.705107e+05
221  B0041   BE-24720       331  1491.368393  1.447646e+03
474  B0032    BE-2572       112    10.416047  2.271066e+06
358  F0033  FR-263690       324  1456.461676  1.150020e+03

In [48]:
sites_df['country'] = sites_df['polyID'].apply(lambda x: 'FR' if x.startswith('FR') else 'BE')
sites_df.sample(10)

Site     polyID  classCLC         dist       surface country
206  B0045   BE-24840       511  1142.543367  2.213737e+05      BE
336  F0029   FR-62513       211   224.912039  3.488307e+06      FR
212  B0041    BE-5289       142  1132.876568  2.070608e+05      BE
127  F0019  FR-143768       242     0.000000  1.235651e+06      FR
380  B0096   BE-19532       243   870.935228  9.364097e+05      BE
67   F0073  FR-256690       324  1061.489087  1.621549e+06      FR
315  F0017  FR-134668       242  1051.046334  1.462918e+06      FR
181  B0163    BE-4067       121   304.285917  4.918914e+05      BE
421  B0019   BE-22542       312     0.000000  5.146666e+06      BE
623  B0158    BE-5779       211  1123.690449  3.326810e+06      BE

In [49]:
sites_df['country'].value_counts()

FR    335
BE    297
Name: country, dtype: int64

In [50]:
sites_df = sites_df.drop(['polyID'], axis=1)
sites_df.sample(5)

Site  classCLC         dist       surface country
210  B0041       121  1169.090590  1.802064e+05      BE
421  B0019       312     0.000000  5.146666e+06      BE
324  F0021       242     0.000000  2.754422e+06      FR
233  B0043       112    58.511637  3.943985e+05      BE
628  B0158       231  1103.848468  4.010538e+05      BE

##### Get one dataset with : beehives and area/distance for each zone

In [51]:
surfaces_df

Site  111           112           121            122  123  124  131  132  \
0   F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1   F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2   F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3   B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4   F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   
..    ...  ...           ...           ...            ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00       0.000000  0.0  0.0  0.0    0   
59  B0069  0.0  8.704243e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
60  B0152  0.0  4.575279e+06  5.726391e+05  201434.000214  0.0  0.0  0.0    0   
61  F0003  0.0  3.425116e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
62  B0158  0.0  1.467398e+05  0.000000e+00  318178.187439  0.0  0.0  0.0    0   

    133  ...  411  412  421  422  423  511            512  521  522  523  
0   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4   0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...            ...  ...  ...  ...  
58  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
59  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
60  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
61  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
62  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  

[63 rows x 45 columns]

In [52]:
classCLC = surfaces_df.columns.to_list()[1:]

In [53]:
by_sites_df = surfaces_df.copy(deep=True)
by_sites_df

Site  111           112           121            122  123  124  131  132  \
0   F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1   F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2   F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3   B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4   F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   
..    ...  ...           ...           ...            ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00       0.000000  0.0  0.0  0.0    0   
59  B0069  0.0  8.704243e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
60  B0152  0.0  4.575279e+06  5.726391e+05  201434.000214  0.0  0.0  0.0    0   
61  F0003  0.0  3.425116e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
62  B0158  0.0  1.467398e+05  0.000000e+00  318178.187439  0.0  0.0  0.0    0   

    133  ...  411  412  421  422  423  511            512  521  522  523  
0   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4   0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...            ...  ...  ...  ...  
58  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
59  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
60  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
61  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
62  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  

[63 rows x 45 columns]

In [54]:
for CLC in classCLC:
    by_sites_df[CLC] = by_sites_df.apply(distance_area, args = (CLC,), axis = 1)

In [55]:
by_sites_df

Site  111           112           121         122  123  124  131  132  \
0   F0069  0.0  1.442780e+03  3.587087e+05    0.000000  0.0  0.0  0.0  0.0   
1   F0025  0.0  0.000000e+00  0.000000e+00  700.801763  0.0  0.0  0.0  0.0   
2   F0027  0.0  2.604161e+02  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
3   B0128  0.0  2.412962e+03  4.053755e+06  523.436355  0.0  0.0  0.0  0.0   
4   F0023  0.0  0.000000e+00  7.907522e+02    0.000000  0.0  0.0  0.0  0.0   
..    ...  ...           ...           ...         ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
59  B0069  0.0  7.328884e+03  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
60  B0152  0.0  4.575279e+06  5.227880e+02  421.402885  0.0  0.0  0.0  0.0   
61  F0003  0.0  3.859675e+02  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
62  B0158  0.0  1.574508e+02  0.000000e+00  392.706036  0.0  0.0  0.0  0.0   

    133  ...  411  412  421  422  423  511         512  521  522  523  
0   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
1   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
2   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
3   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
4   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  463.344615  0.0  0.0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...         ...  ...  ...  ...  
58  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
59  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
60  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
61  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
62  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  

[63 rows x 45 columns]

check if correct

In [56]:
distance = distances_df[(distances_df['Site'] == 'F0023') & 
                        (distances_df['classCLC'] == 512)]['dist'].values
distance

array([614.50674791])

In [57]:
area = surfaces_df[surfaces_df['Site'] == 'F0023']['512'].values[0]
area

284728.392578125

In [58]:
area/distance[0]

463.3446150839585

##### Drop some colomns that don't contain data

In [59]:
by_sites_df = by_sites_df.drop(['132','212','213','222', '223', '241', '244',
              '321', '323', '332', '333', '334', '335', '411',
              '421', '422', '521', '522', '523'], axis=1)

### Merging the datasets into two

3 datasets to 2 : sites + (heavy metal || pesticides)

##### Heavy Metal

In [60]:
hm_df.sample(5)

REF....SUBSTANCE   PERIOD  mrl_lead  mrl_cadmium  mrl_arsenic  mrl_copper  \
321           FE0001  2020/P1         0            0            0           0   
416            B0077  2020/P4         0            0            0           0   
418            B0103  2020/P4         0            0            0           0   
327            B0023  2020/P2         0            0            0           0   
175            B0019  2019/P1         0            0            0           0   

     has_mercury  
321            0  
416            0  
418            0  
327            0  
175            0

In [61]:
group_hm_df.sample(5)

mrl_lead  mrl_cadmium  mrl_arsenic  mrl_copper  has_mercury
REF....SUBSTANCE                                                             
B0067                    0            1            0           0            0
F0017                    0            0            0           0            0
F0031                    0            0            0           0            0
F0001                    0            0            0           0            0
F0005                    0            0            0           0            0

In [62]:
cat_hm_df.sample(5)

cat_label
REF....SUBSTANCE           
B0021                     4
F0073                     0
F0025                     0
B0140                     1
B0039                     0

##### Pesticides

full

In [63]:
pesticides_df.sample(5)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
469            F0025  2020/P3     0.0     0.0000           0.0           0.0   
185            B0093  2018/P4     0.0     0.0000           0.0           0.0   
365            F0013  2020/P1     0.0     0.0407           0.0           0.0   
356            B0140  2020/P1     0.0     0.0000           0.0           0.0   
93             B0142  2018/P1     0.0     0.0000           0.0           0.0   

     Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  \
469           0.0           0.0          0.0        0.0  ...   
185           0.0           0.0          0.0        0.0  ...   
365           0.0           0.0          0.0        0.0  ...   
356           0.0           0.0          0.0        0.0  ...   
93            0.0           0.0          0.0        0.0  ...   

     qtt_insecticides  qtt_herbicides  nb_fungicides  nb_insecticides  \
469               0.0          0.0000              0                0   
185               0.0          0.0146              0                0   
365               0.0          0.0000              1                0   
356               0.0          0.0227              2                0   
93                0.0          0.0000              0                0   

     nb_herbicides  mrl_fungicides  mrl_insecticides  mrl_herbicides  year  \
469              0               0                 0               0  2020   
185              1               0                 0               0  2018   
365              0               0                 0               0  2020   
356              2               0                 0               0  2020   
93               0               0                 0               0  2018   

     trimester  
469          3  
185          4  
365          1  
356          1  
93           1  

[5 rows x 128 columns]

In [64]:
small_pesticides_df.sample(5)

REF....SUBSTANCE  qtt_fungicides  qtt_insecticides  qtt_herbicides  \
502            F0011          0.0000               0.0             0.0   
291            B0096          0.0046               0.0             0.0   
76             B0017          0.0676               0.0             0.0   
364            F0011          0.0000               0.0             0.0   
115            B0043          0.0000               0.0             0.0   

     nb_fungicides  nb_insecticides  nb_herbicides  mrl_fungicides  \
502              0                0              0               0   
291              1                0              0               0   
76               1                0              0               1   
364              0                0              0               0   
115              0                0              0               0   

     mrl_insecticides  mrl_herbicides  year trimester  
502                 0               0  2020         4  
291                 0               0  2019         3  
76                  0               0  2018         1  
364                 0               0  2020         1  
115                 0               0  2018         2

In [65]:
group_pesti_df.sample(5)

mrl_fungicides  mrl_insecticides  mrl_herbicides
REF....SUBSTANCE                                                  
F0073                          1                 1               0
B0017                          1                 0               0
B0061                          0                 0               0
F0069                          0                 0               1
B0105                          1                 1               1

In [66]:
cat_pesti_df.sample(5)

cat_label
REF....SUBSTANCE           
FE0001                    0
B0069                     5
B0017                     3
F0063                     3
F0029                     3

##### sites

In [67]:
sites_df.sample(5)

Site  classCLC         dist       surface country
629  B0158       242  1229.411653  2.034749e+05      BE
182  B0163       122  1295.260111  2.477849e+04      BE
385  B0023       231   301.686590  9.149756e+05      BE
274  B0059       242  1260.189017  1.382229e+06      BE
359  F0033       242    48.493853  7.983372e+05      FR

In [68]:
by_sites_df.sample(5)

Site  111           112           121         122  123  124  131  133  \
3   B0128  0.0   2412.961618  4.053755e+06  523.436355  0.0  0.0  0.0  0.0   
52  B0067  0.0  51710.751583  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
49  F0061  0.0   4997.551880  8.256698e+01    0.000000  0.0  0.0  0.0  0.0   
57  F0011  0.0    753.197885  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
6   F0073  0.0      0.000000  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   

    141  ...          311           312       313          322          324  \
3   0.0  ...     0.000000      0.000000  0.000000     0.000000     0.000000   
52  0.0  ...   372.984311      0.000000  0.000000     0.000000     0.000000   
49  0.0  ...  6298.817449      0.000000  0.000000     0.000000     0.000000   
57  0.0  ...  4990.849187      0.000000  0.000000  2688.189175   289.114682   
6   0.0  ...     0.000000  24566.842848  4.785294     0.000000  1924.689337   

    331  412  423  511  512  
3   0.0  0.0  0.0  0.0  0.0  
52  0.0  0.0  0.0  0.0  0.0  
49  0.0  0.0  0.0  0.0  0.0  
57  0.0  0.0  0.0  0.0  0.0  
6   0.0  0.0  0.0  0.0  0.0  

[5 rows x 26 columns]

##### Merge HM

In [69]:
cat_hm_df.index.name

'REF....SUBSTANCE'

In [70]:
cat_hm_df = cat_hm_df.reset_index()

In [71]:
cat_hm_df = cat_hm_df.rename(columns={"REF....SUBSTANCE": "Site"})

In [72]:
cat_hm_df.head(3)

Site  cat_label
0  B0012          0
1  B0017          0
2  B0019          1

In [73]:
by_sites_df.head(3)

Site  111          112            121         122  123  124  131  133  \
0  F0069  0.0  1442.780424  358708.730957    0.000000  0.0  0.0  0.0  0.0   
1  F0025  0.0     0.000000       0.000000  700.801763  0.0  0.0  0.0  0.0   
2  F0027  0.0   260.416070       0.000000    0.000000  0.0  0.0  0.0  0.0   

   141  ...           311  312  313  322         324  331  412  423  511  512  
0  0.0  ...    352.189664  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  0.0  0.0  
1  0.0  ...  24001.488078  0.0  0.0  0.0  926.217913  0.0  0.0  0.0  0.0  0.0  
2  0.0  ...  11215.516811  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  0.0  0.0  

[3 rows x 26 columns]

In [74]:
merged_hm = pd.merge(by_sites_df, cat_hm_df, on=["Site"])

In [75]:
merged_hm.sample(5)

Site  111          112  121          122         123  124  131  133  141  \
40  B0019  0.0     0.000000  0.0     0.000000    0.000000  0.0  0.0  0.0  0.0   
55  F0007  0.0     0.000000  0.0     0.000000    0.000000  0.0  0.0  0.0  0.0   
21  B0045  0.0  4456.532125  0.0  1032.436781  201.391318  0.0  0.0  0.0  0.0   
34  F0033  0.0    22.019966  0.0     0.000000    0.000000  0.0  0.0  0.0  0.0   
54  F0011  0.0   753.197885  0.0     0.000000    0.000000  0.0  0.0  0.0  0.0   

    ...           312          313          322         324         331  \
40  ...  5.146666e+06  1701.143283     0.000000    0.000000    0.000000   
55  ...  4.060461e+01     0.000000     0.000000    0.000000    0.000000   
21  ...  0.000000e+00     0.000000     0.000000    0.000000  741.986804   
34  ...  2.274376e+03  2808.259665   700.999059    0.789598    0.000000   
54  ...  0.000000e+00     0.000000  2688.189175  289.114682    0.000000   

           412         423         511         512  cat_label  
40  818.792332    0.000000    0.000000    0.000000          1  
55    0.000000    0.000000    0.000000    0.000000          0  
21    0.000000  294.730719  193.755188  118.601578          0  
34    0.000000    0.000000    0.000000    0.000000          0  
54    0.000000    0.000000    0.000000    0.000000          0  

[5 rows x 27 columns]

In [76]:
merged_hm.shape

(60, 27)

#### Merge Pesticides

In [77]:
cat_pesti_df.index.name

'REF....SUBSTANCE'

In [78]:
cat_pesti_df = cat_pesti_df.reset_index()

In [79]:
cat_pesti_df = cat_pesti_df.rename(columns={"REF....SUBSTANCE": "Site"})

In [80]:
cat_pesti_df.head(3)

Site  cat_label
0  B0012          3
1  B0017          3
2  B0019          0

In [81]:
by_sites_df.head(3)

Site  111          112            121         122  123  124  131  133  \
0  F0069  0.0  1442.780424  358708.730957    0.000000  0.0  0.0  0.0  0.0   
1  F0025  0.0     0.000000       0.000000  700.801763  0.0  0.0  0.0  0.0   
2  F0027  0.0   260.416070       0.000000    0.000000  0.0  0.0  0.0  0.0   

   141  ...           311  312  313  322         324  331  412  423  511  512  
0  0.0  ...    352.189664  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  0.0  0.0  
1  0.0  ...  24001.488078  0.0  0.0  0.0  926.217913  0.0  0.0  0.0  0.0  0.0  
2  0.0  ...  11215.516811  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  0.0  0.0  

[3 rows x 26 columns]

In [82]:
merged_pesti = pd.merge(by_sites_df, cat_pesti_df, on=["Site"])

In [83]:
merged_pesti.sample(5)

Site          111           112          121          122  123  124  \
38  B0077     0.000000  4.108273e+02     0.000000    87.748165  0.0  0.0   
15  F0065  1967.704980  3.352559e+06  1416.685638  2509.438121  0.0  0.0   
26  F0055     0.000000  1.313065e+03  2656.270015     0.000000  0.0  0.0   
43  F0061     0.000000  4.997552e+03    82.566983     0.000000  0.0  0.0   
19  B0041   391.077976  2.904582e+06   154.142379     0.000000  0.0  0.0   

           131  133         141  ...  312  313  322          324       331  \
38    0.000000  0.0    0.000000  ...  0.0  0.0  0.0     0.000000  0.000000   
15    0.000000  0.0  191.539631  ...  0.0  0.0  0.0     0.000000  0.000000   
26  121.787548  0.0    0.000000  ...  0.0  0.0  0.0  2812.564594  0.000000   
43    0.000000  0.0    0.000000  ...  0.0  0.0  0.0     0.000000  0.000000   
19    0.000000  0.0    0.000000  ...  0.0  0.0  0.0     0.000000  0.970683   

    412  423  511  512  cat_label  
38  0.0  0.0  0.0  0.0          5  
15  0.0  0.0  0.0  0.0          0  
26  0.0  0.0  0.0  0.0          3  
43  0.0  0.0  0.0  0.0          3  
19  0.0  0.0  0.0  0.0          3  

[5 rows x 27 columns]

In [84]:
merged_pesti.shape

(57, 27)

##### Merge All

In [85]:
group_hm_df.index.name

'REF....SUBSTANCE'

In [86]:
group_pesti_df.index.name

'REF....SUBSTANCE'

In [87]:
group_hm_df = group_hm_df.reset_index()
group_pesti_df = group_pesti_df.reset_index()

In [88]:
group_hm_df = group_hm_df.rename(columns={"REF....SUBSTANCE": "Site"})
group_pesti_df = group_pesti_df.rename(columns={"REF....SUBSTANCE": "Site"})

In [89]:
by_sites_df.head(3)

Site  111          112            121         122  123  124  131  133  \
0  F0069  0.0  1442.780424  358708.730957    0.000000  0.0  0.0  0.0  0.0   
1  F0025  0.0     0.000000       0.000000  700.801763  0.0  0.0  0.0  0.0   
2  F0027  0.0   260.416070       0.000000    0.000000  0.0  0.0  0.0  0.0   

   141  ...           311  312  313  322         324  331  412  423  511  512  
0  0.0  ...    352.189664  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  0.0  0.0  
1  0.0  ...  24001.488078  0.0  0.0  0.0  926.217913  0.0  0.0  0.0  0.0  0.0  
2  0.0  ...  11215.516811  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  0.0  0.0  

[3 rows x 26 columns]

In [90]:
group_hm_df.head(3)

Site  mrl_lead  mrl_cadmium  mrl_arsenic  mrl_copper  has_mercury
0  B0012         0            0            0           0            0
1  B0017         0            0            0           0            0
2  B0019         0            0            0           0            1

In [91]:
group_pesti_df.head(3)

Site  mrl_fungicides  mrl_insecticides  mrl_herbicides
0  B0012               1                 0               0
1  B0017               1                 0               0
2  B0019               0                 0               0

In [92]:
cat_hm_df = cat_hm_df.rename(columns={"cat_label": "hm_cat_label"})
cat_hm_df.head(3)

Site  hm_cat_label
0  B0012             0
1  B0017             0
2  B0019             1

In [93]:
cat_pesti_df = cat_pesti_df.rename(columns={"cat_label": "pesti_cat_label"})
cat_pesti_df.head(3)

Site  pesti_cat_label
0  B0012                3
1  B0017                3
2  B0019                0

In [94]:
merged_all = pd.merge(by_sites_df, group_hm_df, on=["Site"])
merged_all = pd.merge(merged_all, cat_hm_df, on=["Site"])
merged_all = pd.merge(merged_all, group_pesti_df, on=["Site"])
merged_all = pd.merge(merged_all, cat_pesti_df, on=["Site"])

In [95]:
merged_all.shape

(57, 36)

### Final Cleaning

##### HM

In [96]:
merged_hm = merged_hm.drop(['Site'], axis=1)

In [97]:
merged_pesti.sample(5)

Site  111          112          121          122  123          124  \
14  F0043  0.0  1502.358180   833.366065     0.000000  0.0  1557.870715   
8   B0142  0.0  6808.837598  6009.480699  1078.987269  0.0     0.000000   
44  F0063  0.0   339.909994     0.000000     0.000000  0.0     0.000000   
37  B0019  0.0     0.000000     0.000000     0.000000  0.0     0.000000   
40  B0103  0.0   440.186508     0.000000     0.000000  0.0     0.000000   

             131  133  141  ...           312          313  322  324  331  \
14  0.000000e+00  0.0  0.0  ...  0.000000e+00     0.000000  0.0  0.0  0.0   
8   0.000000e+00  0.0  0.0  ...  0.000000e+00     0.000000  0.0  0.0  0.0   
44  0.000000e+00  0.0  0.0  ...  0.000000e+00     0.000000  0.0  0.0  0.0   
37  0.000000e+00  0.0  0.0  ...  5.146666e+06  1701.143283  0.0  0.0  0.0   
40  1.160708e+06  0.0  0.0  ...  7.545125e+02  2719.364725  0.0  0.0  0.0   

           412  423  511  512  cat_label  
14    0.000000  0.0  0.0  0.0          0  
8     0.000000  0.0  0.0  0.0          3  
44    0.000000  0.0  0.0  0.0          3  
37  818.792332  0.0  0.0  0.0          0  
40    0.000000  0.0  0.0  0.0          3  

[5 rows x 27 columns]

In [98]:
merged_hm.to_csv('merged_hm.csv')

##### Pesticides

In [99]:
merged_pesti = merged_pesti.drop(['Site'], axis=1)

In [100]:
merged_pesti.sample(5)

111           112            121         122  123  124         131  \
55  0.0  3.859675e+02       0.000000    0.000000  0.0  0.0    0.000000   
11  0.0  3.367678e+03     427.687419  334.489944  0.0  0.0    0.000000   
30  0.0  2.793964e+02       0.000000    0.000000  0.0  0.0  751.937308   
0   0.0  1.442780e+03  358708.730957    0.000000  0.0  0.0    0.000000   
33  0.0  5.206697e+06       0.000000    0.000000  0.0  0.0    0.000000   

           133  141           142  ...         312          313  322  324  \
55    0.000000  0.0  0.000000e+00  ...  948.728456  5391.035821  0.0  0.0   
11    0.000000  0.0  4.940313e+03  ...    0.000000     0.000000  0.0  0.0   
30  338.605936  0.0  1.560227e+06  ...    0.000000     0.000000  0.0  0.0   
0     0.000000  0.0  1.742440e+02  ...    0.000000     0.000000  0.0  0.0   
33    0.000000  0.0  0.000000e+00  ...    0.000000   178.204219  0.0  0.0   

    331  412  423  511  512  cat_label  
55  0.0  0.0  0.0  0.0  0.0          3  
11  0.0  0.0  0.0  0.0  0.0          3  
30  0.0  0.0  0.0  0.0  0.0          3  
0   0.0  0.0  0.0  0.0  0.0          1  
33  0.0  0.0  0.0  0.0  0.0          0  

[5 rows x 26 columns]

In [101]:
merged_pesti.to_csv('merged_pesti.csv')

##### All

In [102]:
merged_all = merged_all.drop(['Site'], axis=1)

In [103]:
merged_all.sample(5)

111           112           121          122  123  124  \
13     0.000000  0.000000e+00  3.383666e+06   979.946280  0.0  0.0   
40     0.000000  4.401865e+02  0.000000e+00     0.000000  0.0  0.0   
10   209.227976  6.653099e+03  1.764701e+04     0.000000  0.0  0.0   
54     0.000000  4.575279e+06  5.227880e+02   421.402885  0.0  0.0   
15  1967.704980  3.352559e+06  1.416686e+03  2509.438121  0.0  0.0   

             131  133         141        142  ...  mrl_lead  mrl_cadmium  \
13  0.000000e+00  0.0  506.157789  19.198045  ...         0            0   
40  1.160708e+06  0.0    0.000000  21.528232  ...         0            0   
10  0.000000e+00  0.0    0.000000   0.000000  ...         0            0   
54  0.000000e+00  0.0    0.000000   0.000000  ...         0            0   
15  0.000000e+00  0.0  191.539631   0.000000  ...         0            0   

    mrl_arsenic  mrl_copper  has_mercury  hm_cat_label  mrl_fungicides  \
13            0           0            0             0               0   
40            0           0            1             1               1   
10            0           0            0             0               1   
54            0           0            0             0               1   
15            0           1            0             2               0   

    mrl_insecticides  mrl_herbicides  pesti_cat_label  
13                 0               0                0  
40                 0               0                3  
10                 0               0                3  
54                 0               0                3  
15                 0               0                0  

[5 rows x 35 columns]

In [104]:
merged_all.to_csv('merged_all.csv')

# Questions to ask

- When NaN => information irrelevant or not analysed?
- Difference France/Belgium?
- Is the fact that an insecticide is unauthorized relevant? (=> not the same law in France/Belgium)
- Or do we have to show that even banned pesticides are still found?
- How do we treat other categorical pesticides?
- How to increase the number of datas?

### Bo-bo-bonus

- FR  :  36 beehives for 335 zones
- BE  :  27 beehives for 297 zones

Here the CLC zone are left separate, could group some to test